packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -222,10 +225,81 @@ class UpgradeCommandRunner {
     }
   }
 
+  /// Checks if the Flutter git repository is tracking a "standard remote".
+  ///
+  /// Using `flutter upgrade` is not supported from a non-standard remote. A git
+  /// remote should have the same url as [_flutterGit] to be considered as a
+  /// "standard" remote.
+  ///
+  /// Exits tool if the tracking remote is not standard.
+  void verifyStandardRemote(FlutterVersion localVersion) {
+    // If localVersion.repositoryUrl is null, exit
+    if (localVersion.repositoryUrl == null) {
+      throwToolExit(
+        'Unable to upgrade Flutter: The Flutter SDK is tracking an "unknown" '
+        'remote.\n'
+        'Re-install Flutter by going to '
+        'https://flutter.dev/docs/get-started/install. Alternatively, use "git" '
+        'directly to configure a valid upstream for the branch and retry.'
+      );
+    }
+
+    // Strip `.git` suffix from repository url and _flutterGit
+    final String trackingUrl = _stripDotGit(localVersion.repositoryUrl);
+    final String flutterGitUrl = _stripDotGit(_flutterGit);
+
+    // Exempt the flutter GitHub git SSH remote from this check
+    if (trackingUrl == 'git@github.com:flutter/flutter') {
+      return;
+    }
+
+    if (trackingUrl != flutterGitUrl) {
+      if (globals.platform.environment.containsKey('FLUTTER_GIT_URL')) {
+        // If `FLUTTER_GIT_URL` is set, inform the user to either remove the
+        // `FLUTTER_GIT_URL` environment variable or set it to the current
+        // tracking remote to continue.
+        throwToolExit(
+          'Unable to upgrade Flutter: The Flutter SDK is tracking '
+          '"${localVersion.repositoryUrl}" but "FLUTTER_GIT_URL" is set to '
+          '"$_flutterGit".\n'
+          'Either remove "FLUTTER_GIT_URL" from the environment or set '
+          '"FLUTTER_GIT_URL" to "${localVersion.repositoryUrl}", and retry. '
+          'Alternatively, re-install Flutter by going to '
+          'https://flutter.dev/docs/get-started/install.\n'
+          'If this is intentional, it is recommended to use "git" directly to '
+          'keep Flutter SDK up-to date.'
+        );
+      } else {
+        // Inform that the user has to set the environment variable to continue.
+        throwToolExit(
+          'Unable to upgrade Flutter: The Flutter SDK is tracking a non-standard '
+          'remote "${localVersion.repositoryUrl}".\n'
+          'Set the environment variable "FLUTTER_GIT_URL" to '
+          '"${localVersion.repositoryUrl}", and retry. '
+          'Alternatively, re-install Flutter by going to '
+          'https://flutter.dev/docs/get-started/install.\n'
+          'If this is intentional, it is recommended to use "git" directly to '
+          'keep Flutter SDK up-to date.'
+        );
+      }
+    }
+  }
+
+  String _stripDotGit(String url) {
+    final RegExp pattern = RegExp(r'(.*)(\.git)$');
```

Comment:
nit: document the pattern this is intended to match with some examples

Reply:
Done.